net/libfab/libfab.c

```diff
@@ -0,0 +1,525 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * For any questions about this software or licensing,
+ * please email opensource@seagate.com or cortx-questions@seagate.com.
+ *
+ */
+
+
+/**
+ * Mapping for m0net-->libfabric API
+ * xo_dom_init                    =  fi_getinfo, fi_fabric, fi_domain
+ * xo_dom_fini                    =  fi_close, // free resources
+ * xo_tm_init                     =  // Empty function ? fi_send
+ * xo_tm_confine                  =
+ * xo_tm_start                    =  // fi_send
+ * xo_tm_stop                     =  fi_cancel
+ * xo_tm_fini                     =  // Empty function
+ * xo_end_point_create            =  fi_endpoint, fi_pep, fi_av, fi_cq, fi_cntr, fi_eq, fi_bind(av/cq/cntr/eq), fi_pep_bind
+ * xo_buf_register                =  fi_mr_reg, fi_mr_desc, fi_mr_key, fi_mr_bind, fi_mr_enable
+ * xo_buf_deregister              =  fi_close
+ * xo_buf_add                     =  fi_send/fi_recv
+ * xo_buf_del                     =  fi_cancel 
+ * xo_bev_deliver_sync            =  Is it needed?
+ * xo_bev_deliver_all             =  Is it needed?
+ * xo_bev_pending                 =  Is it needed?
+ * xo_bev_notify                  =  Is it needed?
+ * xo_get_max_buffer_size         =  // need to define new API
+ * xo_get_max_buffer_segment_size =  // need to define new functions
+ * xo_get_max_buffer_segments     =  // need to define new functions
+ * xo_get_max_buffer_desc_size    =  
+ * 
+ */
+
+/**
+ * @addtogroup netlibfab
+ *
+ * @{
+ */
+
+#define M0_TRACE_SUBSYSTEM M0_TRACE_SUBSYS_NET
+#include "lib/trace.h"          /* M0_ENTRY() */
+#include "net/net.h"            /* struct m0_net_domain */
+#include "lib/memory.h"         /* M0_ALLOC_PTR()*/
+#include "libfab_internal.h"    /* struct m0_fab__dom_param */
+
+/** Parameters required for libfabric configuration */
+enum m0_fab__mr_params {
+	/** Fabric memory access. */
+	FAB_MR_ACCESS  = (FI_READ | FI_WRITE | FI_RECV | FI_SEND | \
+			FI_REMOTE_READ | FI_REMOTE_WRITE),
+	/** Fabric memory offset. */
+	FAB_MR_OFFSET  = 0,
+	/** Fabric memory flag. */
+	FAB_MR_FLAG    = 0,
+	/** Key used for memory registration. */
+	FAB_MR_KEY     = 0XABCD,
+};
+
+static int libfab_ep_addr_decode(const char *ep_name, char *node, char *port);
+
+/* Used to take the ip and port from the given end point */
+static int libfab_ep_addr_decode(const char *ep_name, char *node, char *port)
+{
+	char     *cp = strchr(ep_name, ':');
+	size_t    n  = cp - ep_name;
+	int       rc = 0;
+
+	M0_PRE(ep_name != NULL);
+
+	if (cp == NULL || n == 0 )
+		return M0_ERR(-EINVAL);
+
+	M0_ENTRY("ep_name=%s", ep_name);
+
+	memcpy(node, ep_name, n);
```

Comment:
How do we make sure the 'node' buffer is large enough?

Reply:
I think we should also pass buffer length as an argument and pre validate if it is large enough